high_card.rb

```diff
@@ -1,5 +1,104 @@
-# Use these two arrays to generate a deck of cards.
-ranks = ["A", 2, 3, 4, 5, 6, 7, 8, 9, 10, "J", "Q", "K" ]
-suits = [ "hearts", "spades", "clubs", "diamonds" ]
+require "pry"
 
-players = []
+# Contains the requred logic for a deck of cards.
+class Deck
+  @@ranks = [2, 3, 4, 5, 6, 7, 8, 9, 10, "J", "Q", "K", "A" ]
+  @@suits = [ "hearts", "spades", "clubs", "diamonds" ]
+  attr_reader :cards
+  # Fills @cards with cards designated by @@ranks, and @@suits
+  # The cards total value is higher as you reache the end of both
+  # arrays.  E.G. A of diamonds is the highest value card, 2 of hearts
+  # the lowest.
+  def initialize
+    @cards = []
+    @@suits.each_with_index{|s, i|
+      totalValue = 0
+      @@ranks.each{|r|
+        cards << {rank: r.to_s, suit: s, totalValue: i+totalValue}
+        totalValue += @@suits.length
+      }
+    }
+    @cards.shuffle!
+  end
+  # Removes the top card from the deck and returns it.
+  def take_card
+    return cards.pop
+  end
+  #  Returns the number of cards left in the deck
+  def num_remaining
+    return cards.length
+  end
+end
+
+# Organizes player data
+class Player
+  attr_accessor :hand, :name
+  def initialize (name)
+    @name = name
+    @hand = nil
+  end
+end
+
+# Organizes game data.  All games have a collection of players and a deck.
+class Game
+  attr_reader :players, :deck
+  # Creates an empty collection of players and a new deck.
+  def initialize
+    @deck = Deck.new
+    @players = []
+  end
+  # Add a player to the game
+  def add_player name
+    players << Player.new(name)
+  end
+  # Deals a single card for each player in the game.
+  def deal_cards
+    # If there's not enough cards to cover all the remaining players create a new deck
+    # and notify the players.
+    if deck.num_remaining < players.length
+      puts "Not enough cards in deck to cover remaining players. Building new deck."
+      @deck = Deck.new # Why do I need @ for this to work?
```

Comment:
instance variables are available throughout the class where as local vars in ruby are only local to their method's scope.